File: core/event-system.js
// Bucu Core - Event System (JavaScript)
// Pub-sub event system with error isolation and rate limiting

const EventSystem = {
    _listeners: new Map(),     // Map<string, Function[]>
    _rateLimits: new Map(),    // Map<string, { count: number, resetTime: number }>
    _stats: {
        emitted: 0,
        blocked: 0,
        errors: 0
    },
    
    // Register an event listener
    on(eventName, callback) {
        // Validate inputs
        if (typeof eventName !== "string" || eventName === "") {
            if (typeof Logger !== 'undefined') {
                Logger.error("Event name must be a non-empty string");
            }
            return false;
        }
        
        if (typeof callback !== "function") {
            if (typeof Logger !== 'undefined') {
                Logger.error("Callback must be a function");
            }
            return false;
        }
        
        // Initialize listener array if needed
        if (!this._listeners.has(eventName)) {
            this._listeners.set(eventName, []);
        }
        
        // Add callback to listeners
        this._listeners.get(eventName).push(callback);
        
        if (typeof Logger !== 'undefined') {
            Logger.debug(`Event listener registered: ${eventName}`);
        }
        
        return true;
    },
    
    // Emit an event
    emit(eventName, data) {
        // Validate event name
        if (typeof eventName !== "string" || eventName === "") {
            if (typeof Logger !== 'undefined') {
                Logger.error("Event name must be a non-empty string");
            }
            return false;
        }
        
        // Check rate limit
        if (!this._checkRateLimit(eventName)) {
            this._stats.blocked++;
            if (typeof Logger !== 'undefined') {
                Logger.warn(`Rate limit exceeded for event: ${eventName}`);
            }
            return false;
        }
        
        // Get listeners for this event
        const listeners = this._listeners.get(eventName);
        if (!listeners || listeners.length === 0) {
            // No listeners, but not an error
            return true;
        }
        
        // Invoke all listeners with error isolation
        let successCount = 0;
        let errorCount = 0;
        
        listeners.forEach((callback, index) => {
            try {
                callback(data);
                successCount++;
            } catch (err) {
                errorCount++;
                this._stats.errors++;
                
                if (typeof Logger !== 'undefined') {
                    Logger.error(
                        `Error in event '${eventName}' callback #${index + 1}: ${err.message}`
                    );
                }
            }
        });
        
        this._stats.emitted++;
        
        if (typeof Logger !== 'undefined' && errorCount > 0) {
            Logger.debug(
                `Event '${eventName}' completed: ${successCount} success, ${errorCount} errors`
            );
        }
        
        return true;
    },
    
    // Remove an event listener
    off(eventName, callback) {
        if (typeof eventName !== "string" || eventName === "") {
            return false;
        }
        
        const listeners = this._listeners.get(eventName);
        if (!listeners) {
            return false;
        }
        
        // Find and remove the callback
        const index = listeners.indexOf(callback);
        if (index !== -1) {
            listeners.splice(index, 1);
            
            if (typeof Logger !== 'undefined') {
                Logger.debug(`Event listener removed: ${eventName}`);
            }
            
            return true;
        }
        
        return false;
    },
    
    // Remove all listeners for an event
    removeAllListeners(eventName) {
        if (eventName) {
            this._listeners.delete(eventName);
            if (typeof Logger !== 'undefined') {
                Logger.debug(`All listeners removed for event: ${eventName}`);
            }
        } else {
            this._listeners.clear();
            if (typeof Logger !== 'undefined') {
                Logger.debug("All event listeners removed");
            }
        }
    },
    
    // Get listener count for an event
    listenerCount(eventName) {
        const listeners = this._listeners.get(eventName);
        return listeners ? listeners.length : 0;
    },
    
    // Get all registered event names
    eventNames() {
        return Array.from(this._listeners.keys());
    },
    
    // Check rate limit for an event
    _checkRateLimit(eventName) {
        // Get rate limit config
        const rateLimitEnabled = typeof ConfigManager !== 'undefined' 
            ? ConfigManager.get("rateLimit.enabled", true)
            : true;
        
        if (!rateLimitEnabled) {
            return true;
        }
        
        // Get event-specific or default limit
        let limit = typeof ConfigManager !== 'undefined'
            ? ConfigManager.get(`rateLimit.events.${eventName}.limit`)
            : null;
        
        let window = typeof ConfigManager !== 'undefined'
            ? ConfigManager.get(`rateLimit.events.${eventName}.window`)
            : null;
        
        if (!limit) {
            limit = typeof ConfigManager !== 'undefined'
                ? ConfigManager.get("rateLimit.defaultLimit", 100)
                : 100;
            window = typeof ConfigManager !== 'undefined'
                ? ConfigManager.get("rateLimit.window", 60)
                : 60;
        }
        
        // Initialize rate limit tracking
        if (!this._rateLimits.has(eventName)) {
            this._rateLimits.set(eventName, {
                count: 0,
                resetTime: Date.now() + (window * 1000)
            });
        }
        
        const rateLimit = this._rateLimits.get(eventName);
        const now = Date.now();
        
        // Reset counter if window expired
        if (now >= rateLimit.resetTime) {
            rateLimit.count = 0;
            rateLimit.resetTime = now + (window * 1000);
        }
        
        // Check if limit exceeded
        if (rateLimit.count >= limit) {
            return false;
        }
        
        // Increment counter
        rateLimit.count++;
        return true;
    },
    
    // Get event system statistics
    getStats() {
        let listenerCount = 0;
        for (const listeners of this._listeners.values()) {
            listenerCount += listeners.length;
        }
        
        return {
            events: this._listeners.size,
            listeners: listenerCount,
            emitted: this._stats.emitted,
            blocked: this._stats.blocked,
            errors: this._stats.errors
        };
    },
    
    // Reset rate limits (for testing/debugging)
    resetRateLimits() {
        this._rateLimits.clear();
        if (typeof Logger !== 'undefined') {
            Logger.debug("Rate limits reset");
        }
    }
};

// Export for Node.js and FiveM
if (typeof module !== 'undefined' && module.exports) {
    module.exports = EventSystem;
}

// Make available globally in FiveM
if (typeof global !== 'undefined') {
    global.EventSystem = EventSystem;
}

File: core/event-system.lua
-- Bucu Core - Event System
-- Pub-sub event system with error isolation and rate limiting

local EventSystem = {
    _listeners = {},     -- { [eventName] = { callback1, callback2, ... } }
    _rateLimits = {},    -- { [eventName] = { count, resetTime } }
    _stats = {
        emitted = 0,
        blocked = 0,
        errors = 0
    }
}

-- Register an event listener
function EventSystem:on(eventName, callback)
    -- Validate inputs
    if type(eventName) ~= "string" or eventName == "" then
        if Logger then
            Logger:error("Event name must be a non-empty string")
        end
        return false
    end
    
    if type(callback) ~= "function" then
        if Logger then
            Logger:error("Callback must be a function")
        end
        return false
    end
    
    -- Initialize listener array if needed
    if not self._listeners[eventName] then
        self._listeners[eventName] = {}
    end
    
    -- Add callback to listeners
    table.insert(self._listeners[eventName], callback)
    
    if Logger then
        Logger:debug("Event listener registered: " .. eventName)
    end
    
    return true
end

-- Emit an event
function EventSystem:emit(eventName, data)
    -- Validate event name
    if type(eventName) ~= "string" or eventName == "" then
        if Logger then
            Logger:error("Event name must be a non-empty string")
        end
        return false
    end
    
    -- Check rate limit
    if not self:_checkRateLimit(eventName) then
        self._stats.blocked = self._stats.blocked + 1
        if Logger then
            Logger:warn("Rate limit exceeded for event: " .. eventName)
        end
        return false
    end
    
    -- Get listeners for this event
    local listeners = self._listeners[eventName]
    if not listeners or #listeners == 0 then
        -- No listeners, but not an error
        return true
    end
    
    -- Invoke all listeners with error isolation
    local successCount = 0
    local errorCount = 0
    
    for i, callback in ipairs(listeners) do
        local success, err = pcall(callback, data)
        
        if success then
            successCount = successCount + 1
        else
            errorCount = errorCount + 1
            self._stats.errors = self._stats.errors + 1
            
            if Logger then
                Logger:error(string.format(
                    "Error in event '%s' callback #%d: %s",
                    eventName,
                    i,
                    tostring(err)
                ))
            end
        end
    end
    
    self._stats.emitted = self._stats.emitted + 1
    
    if Logger and errorCount > 0 then
        Logger:debug(string.format(
            "Event '%s' completed: %d success, %d errors",
            eventName,
            successCount,
            errorCount
        ))
    end
    
    return true
end

-- Remove an event listener
function EventSystem:off(eventName, callback)
    if type(eventName) ~= "string" or eventName == "" then
        return false
    end
    
    local listeners = self._listeners[eventName]
    if not listeners then
        return false
    end
    
    -- Find and remove the callback
    for i, cb in ipairs(listeners) do
        if cb == callback then
            table.remove(listeners, i)
            
            if Logger then
                Logger:debug("Event listener removed: " .. eventName)
            end
            
            return true
        end
    end
    
    return false
end

-- Remove all listeners for an event
function EventSystem:removeAllListeners(eventName)
    if eventName then
        self._listeners[eventName] = nil
        if Logger then
            Logger:debug("All listeners removed for event: " .. eventName)
        end
    else
        self._listeners = {}
        if Logger then
            Logger:debug("All event listeners removed")
        end
    end
end

-- Get listener count for an event
function EventSystem:listenerCount(eventName)
    local listeners = self._listeners[eventName]
    return listeners and #listeners or 0
end

-- Get all registered event names
function EventSystem:eventNames()
    local names = {}
    for name, _ in pairs(self._listeners) do
        table.insert(names, name)
    end
    return names
end

-- Check rate limit for an event
function EventSystem:_checkRateLimit(eventName)
    -- Get rate limit config
    local rateLimitEnabled = ConfigManager and ConfigManager:get("rateLimit.enabled", true)
    if not rateLimitEnabled then
        return true
    end
    
    -- Get event-specific or default limit
    local limit = ConfigManager and ConfigManager:get("rateLimit.events." .. eventName .. ".limit")
    local window = ConfigManager and ConfigManager:get("rateLimit.events." .. eventName .. ".window")
    
    if not limit then
        limit = ConfigManager and ConfigManager:get("rateLimit.defaultLimit", 100)
        window = ConfigManager and ConfigManager:get("rateLimit.window", 60)
    end
    
    -- Initialize rate limit tracking
    if not self._rateLimits[eventName] then
        self._rateLimits[eventName] = {
            count = 0,
            resetTime = os.time() + window
        }
    end
    
    local rateLimit = self._rateLimits[eventName]
    local now = os.time()
    
    -- Reset counter if window expired
    if now >= rateLimit.resetTime then
        rateLimit.count = 0
        rateLimit.resetTime = now + window
    end
    
    -- Check if limit exceeded
    if rateLimit.count >= limit then
        return false
    end
    
    -- Increment counter
    rateLimit.count = rateLimit.count + 1
    return true
end

-- Get event system statistics
function EventSystem:getStats()
    local listenerCount = 0
    for _, listeners in pairs(self._listeners) do
        listenerCount = listenerCount + #listeners
    end
    
    return {
        events = #self:eventNames(),
        listeners = listenerCount,
        emitted = self._stats.emitted,
        blocked = self._stats.blocked,
        errors = self._stats.errors
    }
end

-- Reset rate limits (for testing/debugging)
function EventSystem:resetRateLimits()
    self._rateLimits = {}
    if Logger then
        Logger:debug("Rate limits reset")
    end
end

return EventSystem
